File: app/src/main/java/powerball/apps/jacs/powerball/PowerballFragment.java
package powerball.apps.jacs.powerball;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonArrayRequest;
import com.android.volley.toolbox.Volley;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;


public class PowerballFragment extends Fragment {
    public ArrayList<WinningTicket> tickets = new ArrayList<>();
    public SimpleDateFormat inputFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
    public Calendar now,past;
    public RequestQueue requestQueue;
    public int counter = 0;
    public PowerballFragment() {
        // Required empty public constructor
    }

    public static PowerballFragment newInstance() {
        PowerballFragment fragment = new PowerballFragment();

        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_powerball, container, false);
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

    }

    @Override
    public void onDetach() {
        super.onDetach();

    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setRecyclerView(getActivity());
        Log.d("timer","onActivityCreated");
    }

    @Override
    public void onStart() {
        super.onStart();
        Log.d("timer","onstart");
        //setRecyclerView(getActivity());
    }

    public void setRecyclerView(Context mContext){
        counter = 0;
        tickets = new ArrayList<>();
        RecyclerView rv = (RecyclerView)getView().findViewById(R.id.recycle);
        rv.removeAllViews();
        rv.setHasFixedSize(true);
        rv.addItemDecoration(new DividerItemDecoration(mContext, LinearLayoutManager.VERTICAL));
        LinearLayoutManager llm = new LinearLayoutManager(mContext);
        rv.setLayoutManager(llm);

        final RVAdapter rvAdapter = new RVAdapter(tickets,rv,Constants.POWERBALL);
        rv.setAdapter(rvAdapter);

        requestQueue = Volley.newRequestQueue(mContext);
        now = Calendar.getInstance();
        past = Calendar.getInstance();
        past.add(Calendar.MONTH,-1);
        //String url = "http://data.ny.gov/resource/d6yy-54nr.json";
        String url = "https://www.powerball.com/api/v1/numbers/powerball?_format=json&min="+inputFormat.format(past.getTime()).replace(" ","%20")+"&max="+inputFormat.format(now.getTime()).replace(" ","%20");
        Log.d("timer",url);

        JsonArrayRequest jsonArrayRequest = new JsonArrayRequest(url, new Response.Listener<JSONArray>() {
            @Override
            public void onResponse(JSONArray response) {
                try {
                    if (response.length() > 0) {
                        Log.d("timer","response"+response.toString());
                        tickets.clear();
                        for (int i = 0; i < response.length(); i++) {
                            JSONObject jsonObject = response.getJSONObject(i);
                            WinningTicket person = new WinningTicket();
                            if (!jsonObject.isNull("field_draw_date")) {
                                person.date = jsonObject.getString("field_draw_date");
                            }
                            if (!jsonObject.isNull("field_winning_numbers")) {
                                person.winningNumber = jsonObject.getString("field_winning_numbers");
                            }
                            if (!jsonObject.isNull("field_multiplier")) {
                                person.multiplier = jsonObject.getString("field_multiplier");
                            }
                            //if(!first){
                            //    WinningTicket tick = SharedPrefHelper.getSharedOBJECT(getApplicationContext(),"ticket");
                            //    if(tick == null || !tick.equals(person)){
                            //        SharedPrefHelper.setSharedOBJECT(getApplicationContext(),"ticket",person);
                            //    }
                            //    first = true;
                            //}
                            tickets.add(counter, person);
                            rvAdapter.notifyItemInserted(counter);
                            counter++;
                        }
                       // rvAdapter.notifyDataSetChanged();

                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                // do something
                Log.d("timer","error : "+error.getLocalizedMessage() + error.networkResponse.statusCode);
            }
        });

        requestQueue.add(jsonArrayRequest);
        rvAdapter.setOnLoadMoreListener(new OnLoadMoreListener() {

            @Override
            public void onLoadMore() {
                Log.d("timer","loading more");
                tickets.add(null);
                rvAdapter.notifyItemInserted(tickets.size() - 1);

                now = (Calendar)past.clone();
                past.add(Calendar.MONTH,-1);
                Log.d("timer","dates:::::: "+ now.getTime().toString() + "  :  " + past.getTime().toString());
                String url = "https://www.powerball.com/api/v1/numbers/powerball?_format=json&min="+inputFormat.format(past.getTime()).replace(" ","%20")+"&max="+inputFormat.format(now.getTime()).replace(" ","%20");
                JsonArrayRequest jsonArrayRequest = new JsonArrayRequest(url, new Response.Listener<JSONArray>() {
                    @Override
                    public void onResponse(JSONArray response) {
                        try {
                            if (response.length() > 0) {
                                Log.d("timer","response"+response.toString());
                                //tickets.clear();
                                tickets.remove(tickets.size() - 1);
                                rvAdapter.notifyItemRemoved(tickets.size());

                                for (int i = 0; i < response.length(); i++) {
                                    JSONObject jsonObject = response.getJSONObject(i);
                                    WinningTicket person = new WinningTicket();
                                    if (!jsonObject.isNull("field_draw_date")) {
                                        person.date = jsonObject.getString("field_draw_date");
                                    }
                                    if (!jsonObject.isNull("field_winning_numbers")) {
                                        person.winningNumber = jsonObject.getString("field_winning_numbers");
                                    }
                                    if (!jsonObject.isNull("field_multiplier")) {
                                        person.multiplier = jsonObject.getString("field_multiplier");
                                    }
                                    //if(!first){
                                    //    WinningTicket tick = SharedPrefHelper.getSharedOBJECT(getApplicationContext(),"ticket");
                                    //    if(tick == null || !tick.equals(person)){
                                    //        SharedPrefHelper.setSharedOBJECT(getApplicationContext(),"ticket",person);
                                    //    }
                                    //    first = true;
                                    //}
                                    //tickets.add( person);
                                    if(!tickets.contains(person)) {
                                        tickets.add(counter, person);
                                        rvAdapter.notifyItemInserted(counter);
                                        counter++;
                                    }
                                }
                               // rvAdapter.notifyDataSetChanged();
                                //contactAdapter.notifyDataSetChanged();
                                rvAdapter.setLoaded();

                            }
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                }, new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        // do something
                        Log.d("timer","error : "+error.getLocalizedMessage() + error.networkResponse.statusCode);
                    }
                });
                requestQueue.add(jsonArrayRequest);
            }
        });
    }
}

File: app/src/main/java/powerball/apps/jacs/powerball/PowerballFragment.kt
/*
 * Author: John Rowan
 * Description: fragment for retrieving past results for the powerball and displaying them along with the results
 * of the users tickets.
 * Anyone may use this file or anything contained in this project for their own personal use.
 */
package powerball.apps.jacs.powerball

import android.content.Context
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.android.volley.RequestQueue
import com.android.volley.toolbox.JsonArrayRequest
import com.android.volley.toolbox.Volley
import com.google.gson.Gson
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONException
import powerball.apps.jacs.powerball.data.power.PowerballData
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*

//https://data.ny.gov/resource/d6yy-54nr.json
class PowerballFragment : Fragment() {
    val url = "https://data.ny.gov/resource/d6yy-54nr.json"
    var tickets = ArrayList<WinningTicket?>()
    var inputFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    lateinit var now: Calendar
    lateinit var past: Calendar
    lateinit var requestQueue: RequestQueue
    var counter = 0
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_powerball, container, false)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
    }

    override fun onDetach() {
        super.onDetach()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        setRecyclerView(activity)
        Log.d("timer", "onActivityCreated")
    }

    override fun onStart() {
        super.onStart()
        Log.d("timer", "onstart")
        //setRecyclerView(getActivity());
    }

    fun setRecyclerView(mContext: Context?) {
        counter = 0
        tickets = ArrayList()
        val rv = requireView().findViewById<View>(R.id.recycle) as RecyclerView
        rv.removeAllViews()
        rv.setHasFixedSize(true)
        rv.addItemDecoration(DividerItemDecoration(mContext, LinearLayoutManager.VERTICAL))
        val llm = LinearLayoutManager(mContext)
        rv.layoutManager = llm
        val rvAdapter = RVAdapter(tickets, rv, Constants.POWERBALL)
        rv.adapter = rvAdapter
        requestQueue = Volley.newRequestQueue(mContext)
        now = Calendar.getInstance()
        past = Calendar.getInstance()
        past.add(Calendar.MONTH, -1)
        //String url = "http://data.ny.gov/resource/d6yy-54nr.json";

        GlobalScope.launch {
            withContext(Dispatchers.IO){
                try{
                    val json = URL(url).readText()
                    Log.d("jsondata", json)
                    val gson = Gson()
                    val powerballData = gson.fromJson(json, PowerballData::class.java)
                    for(ticket in powerballData){
                        val person = WinningTicket()
                        person.date = ticket.draw_date//jsonObject.getString("field_draw_date")
                        person.winningNumber = ticket.winning_numbers
                        person.multiplier = ticket.multiplier
                        tickets.add(counter, person)
                        rv.post{
                            rvAdapter.notifyItemInserted(counter)
                        }

                        counter++
                    }
                }catch(e : Exception){
                    Log.e("exception coroutine", "exception ${e.toString()}")
                }

            }
        }

    }

    companion object {
        fun newInstance(): PowerballFragment {
            return PowerballFragment()
        }
    }
}